src/util/jsonLog.js

```diff
@@ -0,0 +1,89 @@
+// @flow
+
+import stringify from "json-stable-stringify";
+import fs from "fs-extra";
+import * as C from "./combo";
+
+/**
+ * JsonLog tracks and serializes append-only logs of JSON objects.
+ *
+ * At its heart, it's basically a simple wrapper around an array, which
+ * enforces the rule that items may be appended to it, but never removed.
+ *
+ * It also provides serialization logic. While the log is an array of JSON
+ * objects and thus a JSON object itself, we don't want to store it on disk as
+ * a stringified JSON object. That would lose the structure of an append-only
+ * array, as every change would result in a rewrite of a single massive line of
+ * text, and would be inefficient in Git storage, and make for ugly diffs and
+ * merge conflicts.
+ *
+ * Instead, we store each item on its own line in the output format, using
+ * `stringify` on the individual JSON objects. We support // comment syntax,
+ * and the toString and writeJsonLog methods allow you to provide prefix
+ * comments; we recommend putting prefix comments like
+ * ["AUTOGENERATED FILE", "do not manually edit"].
+ *
+ * The JsonLog is always parsed using a Combo.Parser, which ensures type safety
+ * at runtime.
+ */
+export class JsonLog<T: C.JsonObject> {
+  _items: T[];
+
+  constructor() {
+    this._items = [];
+  }
+
+  append(items: $ReadOnlyArray<T>): JsonLog<T> {
```

Comment:
In my alt implementation I ended up switching this argument to `Iterable<T>`. They make commands much easier to write as you can conditionally yield events, rather than do conditional array building.

Reply:
Changed.